### PR TITLE
fix: improve repair feedback for LLM-as-a-Judge validations

### LINKED ISSUE #1247: bug: RepairTemplateStrategy includes only yes/no for the failing LLM-as-a-Judge requirement in the repair request
After "The email should have a salutation" requirement validation failed, the following information is sent to the LLM by RepairTemplateStrategy.  The repair information don't have information which requirement failed.

```
INFO: 📤 Request to LLM:
Write an email to Olivia using the notes following: Olivia helped the lab over the last few weeks by organizing intern events, advertising the speaker series, and handling issues with snack delivery..

Requirements:
  - The email should have a salutation
  - Use ONLY lowercase letters (no capital letters, no proper nouns capitalized)
  - Include a specific sign-off phrase: 'thank you for everything'


[⭐ THIS IS A REPAIR ATTEMPT - Model receives feedback from previous failure ⭐]
The following requirements failed before:
* no
```

### PR DESCRIPTION
**Pull request by akihikokuroda**
# Pull Request

## Issue
Fix: #1247 

## Description
   When LLM-as-a-Judge validation returns binary yes/no answers, use the requirement
   description as the repair reason instead of just 'yes' or 'no'. This applies to all
   semantic validations (check(), req() without validation_fn, etc.) regardless of
   check_only setting.

   Previously, repair feedback would show:
```
     The following requirements failed before:
     * no
```
   Now it shows:
``` 
     The following requirements failed before:
     * The email should have a salutation
```

   This makes repair feedback more actionable and helps models understand what needs
   to be fixed during repair iterations in RepairTemplateStrategy and MultiTurnStrategy.

### Testing
- [ ] Tests added to the respective file if code was changed
- [ ] New code has 100% coverage if code was added
- [ ] Ensure existing tests and github automation passes (a maintainer will kick off the github automation when the rest of the PR is populated)

### Attribution
- [x] AI coding assistants used: claudecode

---

### Adding a new component, requirement, sampling strategy, or tool?

If your PR adds or modifies one of the types below, check the matching box. A checklist of type-specific review items will be posted as a comment.

- [ ] Component
- [ ] Requirement
- [ ] Sampling Strategy
- [ ] Tool

NOTE: Please ensure you have an issue that has been acknowledged by a core contributor and routed you to open a pull request against this repository. Otherwise, please open an issue before continuing with this pull request.

### CODE REVIEW

File: test/core/test_requirement_helpers.py [hunk 1]
```diff
@@ -87,5 +87,98 @@ def test_plain_string_yes():
     assert default_output_to_bool("YES") is True  # type: ignore
 
 
+# --- Binary detection for repair feedback (PR #1248) ---
+
+
+def test_repair_feedback_binary_detection_yes():
+    """Binary 'yes' should trigger description fallback."""
+    judge_output = "yes"
+    should_use_description = judge_output and judge_output.strip().lower() in (
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]: These 9 tests re-implement the production predicate in the test body and assert it against itself — they never import or call `Requirement.validate`, so they pass even if lines 284-289 of `requirement.py` are reverted. That's zero coverage of the change, against the "100% coverage if code was added" checklist item.

The inline copy has also already drifted from production: `validate` wraps with `str(judge_output)` (requirement.py:287) but these don't, so `test_repair_feedback_binary_detection_none` exercises different logic than the real path.

Please replace these with a test that constructs a `Requirement` and asserts `validate(...)` returns `reason == description` for a mocked `"no"` judge output, and `reason == <judge output>` for a detailed answer.

File: mellea/core/requirement.py [hunk 2]
```diff
@@ -276,9 +276,21 @@ async def validate(
...
             await llm_as_a_judge_result.avalue()
 
+            # LLM-as-a-Judge validation often returns only "yes"/"no" because the
+            # prompt asks for binary classification. However, repair strategies
+            # display the reason to guide the model during repair iterations.
+            # A bare "no" is unhelpful; the requirement description is more
+            # actionable (e.g., "The email should have a salutation" vs "no").
+            judge_output = llm_as_a_judge_result.value
+            reason = judge_output
+            if judge_output:
```
Comment (ajbozarth) [inline mellea/core/requirement.py:286]: `description` is typed `str | None`. When a requirement has no description and the judge returns `"no"`, this replaces the informative `"no"` with `None` — strictly worse than before the PR. Guard on the description being present:

```python
reason = judge_output
if judge_output and self.description:
    if str(judge_output).strip().lower() in ("yes", "no"):
        reason = self.description
```

**Review by ajbozarth — CHANGES_REQUESTED** [review]
The core fix is sound — substituting the description for a bare `yes`/`no` makes repair feedback actionable. Two blocking items before merge: one on the tests, one on a `None` edge case in the production logic.

File: mellea/core/requirement.py (hunk 2 above)
Comment (akihikokuroda) [reply mellea/core/requirement.py:286]: LLM as a judge without a description doesn't make sense.  I believe that the check is not necessary.

File: test/core/test_requirement_helpers.py (hunk 1 above)
Comment (akihikokuroda) [reply test/core/test_requirement_helpers.py]: Fix them.  Now tests are testing the changed logic.

Comment (ajbozarth) [reply test/core/test_requirement_helpers.py]: Thanks for tightening these up — but they still don't test the production code. Each test body re-declares the `if judge_output: ... reason = description` block inline and asserts on *that local copy*; nothing imports or calls `Requirement.validate`. Reverting `requirement.py:284-289` leaves all of these green, so they can't catch a regression in the change they're meant to cover.

The substitution lives in `Requirement.validate`, so the test has to drive that method. `backend.generate_from_context` is the seam to mock — there's already an established pattern for this kind of test in `test/stdlib/requirements/test_requirement.py`. Something like:

```python
from unittest.mock import AsyncMock, patch
from mellea.core import ModelOutputThunk, Requirement
from mellea.stdlib.context import ChatContext

async def _reason_for(judge_value: str, description: str) -> str | None:
    ctx = ChatContext().add(ModelOutputThunk("some generated output"))
    req = Requirement(description)
    thunk = ModelOutputThunk(value=judge_value)
    with patch.object(
        backend, "generate_from_context",
        new=AsyncMock(return_value=(thunk, ctx)),
    ):
        result = await req.validate(backend, ctx)
    return result.reason

async def test_validate_binary_no_uses_description():
    assert await _reason_for("no", "Must have a salutation") == "Must have a salutation"

async def test_validate_detailed_answer_preserved():
    r = await _reason_for("The greeting is present", "Must have a salutation")
    assert r == "The greeting is present"
```

That asserts on the real `ValidationResult.reason` coming out of `validate`, so it actually pins the new behavior. (These would fit best in `test/stdlib/requirements/test_requirement.py` alongside the existing `validate` tests — the snippet is illustrative on the backend setup; the key point is driving `validate` rather than a copied predicate.)

On the `requirement.py:286` point — agreed, a description-less LLMaJ requirement is already a degenerate config, so that one's non-blocking. Resolving it.

Comment (akihikokuroda) [reply test/core/test_requirement_helpers.py]: Tests are updated.

File: test/core/test_requirement_helpers.py [hunk 3]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
     assert default_output_to_bool("YES") is True  # type: ignore
 
 
+# --- Binary detection for repair feedback (PR #1248) ---
+# Tests for Requirement.validate() binary detection: bare "yes"/"no" responses
+# are replaced with the requirement description for more actionable repair feedback.
+# @pytest: unit
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]: `# @pytest:` is the collection directive for `docs/examples/`, not `test/` files (auto-collected), so this line has no effect. The marker it names, `unit`, is *"auto-applied by conftest — never write it explicitly"* per AGENTS.md §3 — same reason the `@pytest.mark.unit` decorators below can come off.

```suggestion
```

File: test/core/test_requirement_helpers.py [hunk 4]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
...
+
+@pytest.fixture
+def mock_backend():
+    """Mock backend for testing Requirement.validate."""
+    from unittest.mock import MagicMock
+
+    return MagicMock()
+
+
+@pytest.mark.unit
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]:
```suggestion
```

File: test/core/test_requirement_helpers.py [hunk 5]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
...
+async def test_validate_binary_no_uses_description(mock_backend):
+    """Binary 'no' should fail with requirement description as reason."""
+    result = await _validate_result(
+        mock_backend, "no", "The email should have a salutation"
+    )
+    assert result.as_bool() is False
+    assert result.reason == "The email should have a salutation"
+
+
+@pytest.mark.unit
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]:
```suggestion
```

File: test/core/test_requirement_helpers.py [hunk 6]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
...
+async def test_validate_binary_yes_uses_description(mock_backend):
+    """Binary 'yes' should pass with requirement description as reason."""
+    result = await _validate_result(
+        mock_backend, "yes", "The email should have a salutation"
+    )
+    assert result.as_bool() is True
+    assert result.reason == "The email should have a salutation"
+
+
+@pytest.mark.unit
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]:
```suggestion
```

File: test/core/test_requirement_helpers.py [hunk 7]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
...
+    result = await _validate_result(
+        mock_backend,
+        "Yes, the email contains a proper greeting",
+        "The email should have a salutation",
+    )
+    assert result.as_bool() is True
+    assert result.reason == "Yes, the email contains a proper greeting"
+
+
+@pytest.mark.unit
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]:
```suggestion
```

File: test/core/test_requirement_helpers.py [hunk 8]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
...
+async def test_validate_binary_detection_whitespace_and_case(mock_backend):
+    """Binary detection should handle whitespace and case variation."""
+    result = await _validate_result(mock_backend, "  NO  ", "Check requirement")
+    assert result.as_bool() is False
+    assert result.reason == "Check requirement"
+
+
+@pytest.mark.unit
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]:
```suggestion
```

File: test/core/test_requirement_helpers.py [hunk 9]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
...
+async def test_validate_empty_string_preserves_output(mock_backend):
+    """Empty string should not trigger fallback."""
+    result = await _validate_result(
+        mock_backend, "", "The email should have a salutation"
+    )
+    assert result.as_bool() is False
+    assert result.reason == ""
+
+
+@pytest.mark.unit
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py]:
```suggestion
```

File: test/core/test_requirement_helpers.py [hunk 10]
```diff
@@ -87,5 +87,103 @@ def test_plain_string_yes():
     assert default_output_to_bool("YES") is True  # type: ignore
 
 
+# --- Binary detection for repair feedback (PR #1248) ---
+# Tests for Requirement.validate() binary detection: bare "yes"/"no" responses
+# are replaced with the requirement description for more actionable repair feedback.
+# @pytest: unit
+
+
+async def _validate_result(backend, judge_value: str, description: str):
```
Comment (ajbozarth) [inline test/core/test_requirement_helpers.py:98]: Optional: the `unittest.mock` / `mellea.core` / `mellea.stdlib.context` imports here and in `mock_backend` could hoist to the module top alongside the existing `from mellea.core import ...`. Pure style, not blocking.

**Review by ajbozarth — APPROVED** [review]
Tests now drive `Requirement.validate` with a mocked backend and assert on both `as_bool()` and the real `ValidationResult.reason` — verified they pin the change (reverting the substitution in `requirement.py` fails 3 of them). Thanks for the iteration. Two non-blocking convention nits below; fine to merge with or without them.

**Comment by akihikokuroda** [comment]
@ajbozarth Thanks for reviews.  I updated for the suggested changes.

**Review by jakelorocco — APPROVED** [review]
